compiler/natives/src/net/http: Implement http.RoundTripper on Fetch API.

shurcooL opened the pull request:
This is a WIP PR in order to get initial review comments, to communicate and avoid duplicate efforts. Known TODOs are listed in code.

The motivation to use Fetch API is because it's lower level, and allows implementing an http.RoundTripper more efficiently (less conversions between strings and []byte). It also supports streaming body responses efficiently.

References:

- https://fetch.spec.whatwg.org/
- https://streams.spec.whatwg.org/
- http://caniuse.com/#search=Fetch

Tested in stable channel of Chrome, latest Safari, Firefox (developer edition).

Dynamically determine which of XHR, Fetch APIs are available and gracefully fallback to the best available API.

Here's a quick demo, using a local instance of https://http2.golang.org/reqinfo and https://http2.golang.org/clockstream endpoints. I've asked @bradfitz if it's possible to enable CORS on http2.golang.org itself, and if so, you could try this out in a playground.

```Go
func streamingDemo() error {
	resp, err := http.Get("https://localhost:4430/clockstream")
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	_, err = io.Copy(os.Stdout, resp.Body)
	return err
}
```

### Results

(It's a video, click it to view.)

[![image](https://cloud.githubusercontent.com/assets/1924134/15110331/0e3848ac-1595-11e6-8f1b-5ae70a85e644.png)](http://virtivia.com:27080/15kvcw9ib4f2u.mov)

shurcooL commented:
Ok, I've setup an instance of `h2demo` with CORS enabled, so now it's possible to try out this change in the GopherJS Playground:

http://www.gopherjs.org/playground/#/5NfiqqGZCs

Wait for it to load, press run, and wait a few seconds as it compiles the program and runs it!

It will make a request to /reqinfo, then another to /clockstream and copy (streaming) first 1500 bytes before returning.

neelance commented:
Nice!

shurcooL commented on compiler/natives/net/http/fetch.go:
```diff
@@ -0,0 +1,126 @@
+			header := Header{}
+			result.Get("headers").Call("forEach", func(value, key *js.Object) {
+				header[CanonicalHeaderKey(key.String())] = []string{value.String()} // TODO: Support multiple values.
+			})
```
Suggestions for how to do this better are very welcome.

What's a good way to iterate all keys (and values) of a JS `Iterable`?

In JS, one could do something like:

```JS
for (let value of ["a", "b", "c"]) {
    console.log(value);
}
```

But what's a good way in GopherJS?

shurcooL commented:
@neelance I've completed the critical functionality (closing response body, full header support, and sending request body). PTAL.

This should be functionally ready to be merged. There are still some style and minor TODOs to clean up. Also, before merging, this should be squashed into logical commits with clean commit message, and `gopherjs/compiler/natives/fs_vfsdata.go` needs to be regenerated.

neelance commented on compiler/natives/src/net/http/http.go:
```diff
@@ -13,18 +13,29 @@ import (
 
-var DefaultTransport RoundTripper = &XHRTransport{}
+var DefaultTransport = func() RoundTripper {
+	if fetchAPI, streamsAPI := js.Global.Get("fetch"), js.Global.Get("ReadableStream"); fetchAPI != js.Undefined && streamsAPI != js.Undefined {
```
Why the extra variables?

shurcooL replied:
It checks that both `fetch` API and `ReadableStream` is available. The former is needed to make Fetch requests, and the latter is to be able to execute this line successfully:

```Go
Body: &streamReader{stream: result.Get("body").Call("getReader")},
```

For example, Firefox currently supports Fetch API but does not support streaming response bodies. So `result.Get("body")` would be nil, and using Fetch over XHR is not very advantageous.

neelance replied:
I meant this:
```
if js.Global.Get("fetch") != js.Undefined && js.Global.Get("ReadableStream") != js.Undefined {
```

shurcooL replied:
Ah, I overlooked that or didn't think of it. It looks simpler, so I'll do it.

I think I initially tried to pass the variable into the transport, e.g.:

```
if xhrAPI := js.Global.Get("XMLHttpRequest"); xhrAPI != js.Undefined {
	return &XHRTransport{xhrAPI}
}
```

But reverted because it was a bad idea. It meant that zero value was no longer valid, and there's no performance gain in storing `js.Global.Get("XMLHttpRequest")` in a variable.

neelance commented on compiler/natives/src/net/http/fetch.go:
```diff
@@ -0,0 +1,143 @@
+		//       See BufferSource at https://fetch.spec.whatwg.org/#body-mixin.
+		body, err := ioutil.ReadAll(req.Body)
+		if err != nil {
+			req.Body.Close() // RoundTrip must always close the body, including on errors.
```
Can this be ensured via a `defer` at the beginning on the function?

shurcooL replied:
Yes, but I don't think it'd be cleaner. You still have to check that `req.Body` is not `nil` before doing it.

There are no other returns before this section of code (the code above simply deals with setting up the fields needed for the request), so `req.Body.Close()` will always be called.

neelance replied:
K.

shurcooL commented:
In case it's helpful, I've tested the implementation significantly during development using https://github.com/shurcooL/play/blob/master/192/go-js-fetch/http/test/main.go. You can run that suite by `go get -u -d github.com/shurcooL/play/192/go-js-fetch/http/test` and `gopherjs serve`ing it.

neelance commented:
LGTM

shurcooL commented:
Squashed and merged as 00254cbca59bacb672a3698e0490367cfc64b36c and db27c7c470d7404b6b201f82d6fd4821260bd13e (with logically separated changes, clean commit messages).

I did not force-push to the `fetch-transport` branch of this PR, so that we can still see the development history here. Instead, I just added "Closes #454." to the commit message of the squashed commit. But its content is identical as the diff in this PR.

shurcooL commented:
For anyone interested in an updated demo, I've updated the playground to use latest GopherJS compiler and created this simple demo:

http://www.gopherjs.org/playground/#/okdxAaGlfF
